[VideoThumbLoader] Do not replace the whole art map when getting item…

enen92 opened the pull request:
…s from the library

## Description
Follow up to https://github.com/xbmc/xbmc/pull/16444. Since the default artwork is now stored in the artmap (instead of a member variable in `CFileItem`) if `setArt` is used on the ListItem with an art map, the whole map is replaced. Defaulticons were added to the map before getting the available artwork from the library so... make sure we add all the art in the library but do not replace the already existing art (unless it has the same type)

@DaveTBlake or @ksooo for review if possible. Sorry for closing the previous pull request - found a better place for the fix and the job was still in the jenkins queue.

## Motivation and Context
Fixes https://github.com/xbmc/xbmc/issues/17034. @ronie please test for any regressions.

## How Has This Been Tested?
Manually with the db provided in the linked issue.

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

enen92 commented on xbmc/video/VideoThumbLoader.cpp:
```diff
@@ -492,7 +492,9 @@ bool CVideoThumbLoader::LoadItemLookup(CFileItem* pItem)
 void CVideoThumbLoader::SetArt(CFileItem &item, const std::map<std::string, std::string> &artwork)
 {
-  item.SetArt(artwork);
+  for (const auto& art : artwork)
```
This shoud be replaced by `CGUIListItem::AppendArt` as the implementation is exactly the same.

ronie reviewed (approved):
tested and works ok as far as i can tell

DaveTBlake commented:
Only thought @enen92 is there any situation where `CVideoThumbLoader::SetArt` is called and expected to clear the previously added (non-default icon) art before adding new?

enen92 commented:
@DaveTBlake, I tracked all uses of this method in the code:

-1)  called in `bool CVideoThumbLoader::LoadItemCached(CFileItem* pItem)` after getting art from the database. In this case I think it doesn't make sense to clear all art (defaultvideo.png should be there)

-2) called in `bool CVideoThumbLoader::LoadItemLookup(CFileItem* pItem)` getting art from the database. I think in this case it doesn't make sense to replace the art.

-3) called in `bool CVideoThumbLoader::FillLibraryArt(CFileItem &item)`  for videos, which is exactly the case we are trying to solve. Should not replace art.

-4) called in `void CGUIDialogVideoInfo::PlayTrailer()` to set the art for the item and ensure fallbacks. I am unsure it should replace it all or not. But since `Append` will also replace any existing art type I think it won't cause any regression.

DaveTBlake commented:
Ah yes, with a map we won't get duplicate entries even if called twice. But do any of the 4 methods you identified ever get called in a place that wants to remove some previously set art? `Append` will replace the image for an art type, but it won't remove an art type completely. 

Most cases I'm sure it is a matter of appending the rest of the art for an item that has the default icon, so this is just me theorising. I don't know of a situation that reloads an item with changed art  - after user manually picks art from the info dialog perhaps?

enen92 commented:
I see a couple of options @DaveTBlake :

1 - Merge this as is and fix any other problems that might be caused by the change in the other 3 calls later (I doubt they will occur). (Fixes this issue and others we might not be seeing)

2 - Modify the `setArt` method to include a `replace` argument. Pass `true` for all calls listed in my last comment except the case this PR is trying to address (item 3 in my last comment). If problems arise in some of the calls in the future, just pass `false`. (Fixes this issue only and not those we can't see)

3 - Modify setArt to include the same argument but pass always false (similar to 1). (Fixes this issue and others we might not be seeing but leaves the method ready for a future fix in some of the calls)

What do you think it's the way to go? I'd prefer option 3 even if issues are caused later.

DaveTBlake commented:
I'm even happy with 1) @enen92, if there are issues later this discussion is easy to find. It is probably fine, I was just tying to think all the possibilities through.

Although perhaps @rmrector has thoughts on this? Anywhere an item with art loaded could be using `CVideoThumbLoader::SetArt ` to remove an art type? 

It is possibly worth taking a quick look at what happens when a type of art  is manually removed for a video via the info dialog "Choose Art" facility (sorry nothing set up to test that myself at the moment).

rmrector commented:
Can you use [`SetArtFallback`](https://github.com/xbmc/xbmc/blob/master/xbmc/guilib/GUIListItem.h#L91) to set the default icon? Fallback images are used if the main art map does not contain the requested image type, and `SetArtwork(ArtMap)` will only override the default icon if the ArtMap contains an icon.

Fallback images are only set for thumb and fanart, so adding icon to the mix won't get crossed with existing behavior.

rmrector commented:
Looks acceptable to me. Those fallbacks are waaaay out of date and make no sense nowadays anyway, maybe remove this `SetArt` method altogether and each of the callers should use whichever `CGUIListItem` method that makes sense directly.

Ultimately, though, I don't think the default icon belongs in this artmap anyway. It should be a fallback for the skinning engine, not a property of each media item, but the nightmare-spaghetti of `CFileItem` and pals doesn't really leave us a distinction.

enen92 commented:
> Looks acceptable to me. Those fallbacks are waaaay out of date and make no sense nowadays anyway, maybe remove this `SetArt` method altogether and each of the callers should use whichever `CGUIListItem` method that makes sense directly.
> 
> Ultimately, though, I don't think the default icon belongs in this artmap anyway. It should be a fallback for the skinning engine, not a property of each media item, but the nightmare-spaghetti of `CFileItem` and pals doesn't really leave us a distinction.

You mean it should be safe to remove the fallbacks for "thumb" as they should be handled by the skin? Won't this break any skin that doesn't implement the said fallbacks? Are we okay with that?
If yes then I think the best option is to remove the method completely and implement the desired logic in each of the callees

rmrector commented:
There is a good chance that some skins still use thumb when they mean poster - Chrous2 also does.

But this was just a helper for skins originally designed for XBMC 11 Eden to upgrade to new versions.

These fallbacks don't make sense now: 1) thumb is a legitimate artwork type of it's own, and 2) its pert-near impossible to design a screen with a box for the image that can fit a 16:9 thumbnail, a 2:3 poster, or a 200:37 banner that doesn't look stupid without knowing which one the actual image is ([example](https://github.com/xbmc/xbmc/pull/15543) of the result nowadays).

sualfred commented:
I'm not sure if I can follow the complete discussion, but a thumb should be a thumb. No poster or anything. It's pita that we don't know the aspect ratio. It's always causing super annoying 16:9 stretches. I guess I can speak for all skinners that we would kill for a aspect ratio info label for each art sometimes.

In my eyes ListItem.Icon should only reflect the DefaultIcon based on the mediatype or provided by the content provider / addon. Icon is a icon. All other ones can be accessed by ListItem.Art(foo) and have the common aspect ratios of their types.

enen92 commented:
Ok so I will drop that method and move to cfileitem start on the callees

DaveTBlake commented:
>Ultimately, though, I don't think the default icon belongs in this artmap anyway. It should be a fallback for the skinning engine, not a property of each media item, but the nightmare-spaghetti of CFileItem and pals doesn't really leave us a distinction.

So you would not have made the orginal change of https://github.com/xbmc/xbmc/pull/16444 putting the default icon in the artmap with art type "icon" then @rmrector ?

rmrector commented:
Ohh, I guess I was coming at that from a data perspective. It makes sense for the GUI, but `CFileItem` subclasses `CGUIListItem` which means this isn't _just_ for the GUI anymore. The code wants to treat an item as simple data in many cases, and I don't think the _default_ icon as the concept exists now (icons generally derived from the media type) belongs in the simple data. Those are the concern of one specific UI - the skin component - and their appearance is determined by the specific skin.

With that as the starting point, I think #16444 and the fixups needed to make it work basically break out even. Some things were simplified while others were complicated, though the new complications seem to come down to mixing the two usages into one interface.

If (my) wishes were fishes, we would have a full compliment of your `CAlbum`-like simple data buckets for all media items AND use them when working with items as simple data internally, then the _default_ icon can make sense in the GUI-only artmap.

enen92 committed:
[VideoThumbLoader] Remove old SetArt fallbacks. Use appendArt to avoid replacing the whole art map

enen92 commented:
@rmrector @DaveTBlake please recheck whenever you find time. Removed the old `VideoThumbLoader::SetArt`(nonsense art fallbacks) and adjusted the callees instead. It'd be interesting to know why the MusicLibrary behaves differently (it's using setArt - c.f  `bool CVideoThumbLoader::FillLibraryArt(CFileItem &item)`) so maybe the fallbacks are being applied later.

rmrector commented:
The new change looks good.

There are a few other calls to `SetArtFallback` that involve "thumb" falling back to a parent "poster" or "banner" like "tvshow.banner" in "CVideoThumbLoader::FillLibraryArt" and a few in "GUIWindowVideoNav.cpp", if you want to clobber the edges of the "thumb" fallbacks.

> It'd be interesting to know why the MusicLibrary behaves differently (it's using setArt - c.f `bool CVideoThumbLoader::FillLibraryArt(CFileItem &item)`) so maybe the fallbacks are being applied later.

Applying these default icons later would help with most of the fallout of the original PR.

DaveTBlake reviewed (approved):
Hitting the limits of my video art knowledge, but if @rmrector is happy then so am I.

>It'd be interesting to know why the MusicLibrary behaves differently (it's using setArt - c.f bool CVideoThumbLoader::FillLibraryArt(CFileItem &item)) so maybe the fallbacks are being applied later. Or maybe the issue will also reflect on the Music Library and the calls should be also changed to AppendArt.

For music the need for fallback art (letting a song have the album cover or having art of its own) and where it is handled are different from video, so no I don't think it needs similar treatment. But thanks for considering it :)

enen92 commented:
@rmrector this PR tries to address an identified issue. I am happy to remove the fallbacks for tvshows (leaving music as is) but I'd prefer to do it in a future pull request if possible since it is not related to the original issue.
cheers

ksooo commented:
@enen92 for me, this breaks PVR channel icons in PVR info dialog:

correct (before this PR):
![screenshot005](https://user-images.githubusercontent.com/3226626/73135611-4783df80-4044-11ea-9d99-aa2efdc45423.png)

wrong (after this PR):
![screenshot004](https://user-images.githubusercontent.com/3226626/73135617-5074b100-4044-11ea-83c2-5f28683cc2f2.png)

enen92 commented:
Thanks for reporting, I'll take a look.

howie-f commented:
unfortunately this commit seems to break movie-posters in chorus2 and official-kodi-remote (ios)

![IMG_0038](https://user-images.githubusercontent.com/58829855/76556341-9e7d3100-6499-11ea-934d-79527e0d6f79.PNG)

reverting the commit for testing brought back the expected behavior

enen92 commented:
@howie-f can you open an issue for it please? I'll self-assign it.
